src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -768,6 +772,40 @@ void MergeTreeData::MergingParams::check(const StorageInMemoryMetadata & metadat
             throw Exception("Version column " + version_column + " does not exist in table declaration.", ErrorCodes::NO_SUCH_COLUMN_IN_TABLE);
     };
 
+    /// Check that if the is_deleted column is needed, it exists and is of type UInt8. If exist, version column must be defined too but version checks are not done here.
+    auto check_is_deleted_column = [this, & columns](bool is_optional, const std::string & storage)
+    {
+        if (is_deleted_column.empty())
+        {
+            if (is_optional)
+                return;
+
+            throw Exception("Logical error: is_deleted (" + is_deleted_column + ") column for storage " + storage + " is empty", ErrorCodes::LOGICAL_ERROR);
+        }
+        else
+        {
+            if (version_column.empty() && !is_optional)
+                throw Exception("Logical error: Version column (" + version_column + ") for storage " + storage + " is empty while is_deleted (" + is_deleted_column + ") is not", ErrorCodes::LOGICAL_ERROR);
```

Comment:
Tip: when formatting Exceptions like this, it is much more convenient to use the `Exception(errorcode, fmt, ...)` constructor

Reply:
I was just trying to stay consistent with previous Exceptions in `check_sign_column`  and `check_version_column`.
But, I agree that is more readable with the error code first, as it's done in some other parts.